src/twisted/mail/imap4.py

```diff
@@ -5493,33 +5495,35 @@ def encoder(s, errors=None):
 
     @param errors: Policy for handling encoding errors.  Currently ignored.
 
-    @return: C{tuple} of a C{str} giving the encoded bytes and an C{int}
+    @return: C{tuple} of a C{bytes} giving the encoded bytes and an C{int}
         giving the number of code units consumed from the input.
     """
-    r = []
+    r = bytearray()
     _in = []
+    valid_chars = set(map(chr, range(0x20,0x7f))) - {"&"}
     for c in s:
-        if ord(c) in (range(0x20, 0x26) + range(0x27, 0x7f)):
+        if c in valid_chars:
             if _in:
-                r.extend(['&', modified_base64(''.join(_in)), '-'])
+                r += b'&' + modified_base64(''.join(_in)) + b'-'
                 del _in[:]
-            r.append(str(c))
+            r.append(ord(c))
         elif c == '&':
             if _in:
-                r.extend(['&', modified_base64(''.join(_in)), '-'])
+                r += b'&' + modified_base64(''.join(_in)) + b'-'
                 del _in[:]
-            r.append('&-')
+            r += b'&-'
         else:
             _in.append(c)
     if _in:
-        r.extend(['&', modified_base64(''.join(_in)), '-'])
-    return (''.join(r), len(s))
+        r.extend(b'&' + modified_base64(''.join(_in)) + b'-')
+    return (bytes(r), len(s))
```

Comment:
It works just fine in Python3, haven't you seen the tests?

the `r` in `bytes(r)` is a `bytearray`